### DISCUSSION TITLE
Tasmota TM1638 LEDs and Switches not working

### DISCUSSION THREAD

**Discussion by usrrsr** [arendst/Tasmota#18127] [Q&A]
I am using TM1638 module having 8 nos 7 segment display, 8 switches and 8 leds and connected to ESP32 wroom-32d board as per following connection
GPIO16 : TM1638_CLK
GPIO17 : TM1638_DIO
GPIO25 : TM1638_STB
I have flashed esp32 with custom build Program Version12.4.0.2(tasmota32) with following option in user_config_override.h
...

#define USE_DISPLAY
#define USE_DISPLAY_TM1637

#define USE_TM1638 // Add support for TM1638 switches copying Switch1 .. Switch8 (+1k code)
//#define TM1638_USE_AS_BUTTON // Add support for buttons
#define TM1638_USE_AS_SWITCH // Add support for switches (default)
#define TM1638_MAX_DISPLAYS 8 // Add support for power control 8 displays
#define TM1638_MAX_KEYS 8 // Add support for 8 keys
#define TM1638_MAX_LEDS 8 // Add support for 8 leds 

I am successfully tested 8 7segment displays, those are working fine and displaying text as well as number.
However i am unable to use 8 Leds and 8 switches, Neither i am getting virtual leds or relays to control it

Can someone help me out to configure it properly and get LED & switches working ?

Regards,

**Comment by barbudor** [comment]
I've just did the same and get everyhting working
Here are important things I see on logs at boot:
```
00:00:00.002 HDW: ESP32-D0WD 
00:00:00.019 UFS: FlashFS mounted with 304 kB free
00:00:00.030 CFG: Loaded from File, Count 55
00:00:00.040 QPC: Count 1
00:00:00.046 SWT: Add vSwitch1, State 0
00:00:00.047 SWT: Add vSwitch2, State 0
00:00:00.047 SWT: Add vSwitch3, State 0
00:00:00.048 SWT: Add vSwitch4, State 0
00:00:00.059 SWT: Add vSwitch5, State 0
00:00:00.059 SWT: Add vSwitch6, State 0
00:00:00.059 SWT: Add vSwitch7, State 0
00:00:00.070 SWT: Add vSwitch8, State 0
...
00:00:00.147 DSP: TM1638 with 8 digits (type 0)
....
19:51:59.831 MQT: stat/esp32dev2/POWER1 = {"POWER1":"OFF"}
19:51:59.833 MQT: stat/esp32dev2/POWER1 = OFF (retained)
19:51:59.846 MQT: stat/esp32dev2/POWER2 = {"POWER2":"OFF"}
19:51:59.848 MQT: stat/esp32dev2/POWER2 = OFF (retained)
19:51:59.850 MQT: stat/esp32dev2/POWER3 = {"POWER3":"OFF"}
19:51:59.862 MQT: stat/esp32dev2/POWER3 = OFF (retained)
19:51:59.864 MQT: stat/esp32dev2/POWER4 = {"POWER4":"OFF"}
19:51:59.866 MQT: stat/esp32dev2/POWER4 = OFF (retained)
19:51:59.883 MQT: stat/esp32dev2/POWER5 = {"POWER5":"OFF"}
19:51:59.886 MQT: stat/esp32dev2/POWER5 = OFF (retained)
19:51:59.889 MQT: stat/esp32dev2/POWER6 = {"POWER6":"OFF"}
19:51:59.891 MQT: stat/esp32dev2/POWER6 = OFF (retained)
19:51:59.904 MQT: stat/esp32dev2/POWER7 = {"POWER7":"OFF"}
19:51:59.906 MQT: stat/esp32dev2/POWER7 = OFF (retained)
19:51:59.908 MQT: stat/esp32dev2/POWER8 = {"POWER8":"OFF"}
19:51:59.920 MQT: stat/esp32dev2/POWER8 = OFF (retained)
19:51:59.923 MQT: stat/esp32dev2/POWER9 = {"POWER9":"OFF"}
19:51:59.925 MQT: stat/esp32dev2/POWER9 = OFF (retained)
```
The 1st series of log confirm the correct configuration of switches from Switch1 to Switch8 assigned to TM1368 buttons
Then the display driver is enabled
At last, you see the startup state of Power1 to Power8 assigned to the leds
With any switch, the matching LED get on when I press and off when I release due to the default `Switchmode<x> 0` 
Changing the mode to 3 for example allows me to toggle a LED with the matching switch

Could you check
- What you have in the logs at boot
- Double check you have configured the GPIO as TM163**8**_xxx and not as TM1637_xxx which appears first in the list

![image](https://user-images.githubusercontent.com/2996042/223205121-b803db06-78b1-4d1c-baf7-8c947045e71e.png)

![image](https://user-images.githubusercontent.com/2996042/223205210-f30f96db-cf0e-4aee-aff3-b528a02f6f1a.png)

![image](https://user-images.githubusercontent.com/2996042/223205256-939c2952-a9b7-4a04-8682-cda3041aa1dd.png)

**Reply by barbudor** [reply]
@usrrsr 
What is your exact configuration ? 
What do you have configured on top of the TM1638 ?
We have just identified a bug where some conflicts between drivers would get the exact impact you are describing.

Please confirm it is working when only the TM1638 GPIO are configured

**Reply by usrrsr** [reply]
My Exact configuration of user_config_override.h is as follows

#define STA_SSID1         "xxx"             // [Ssid1] Wifi SSID
#define STA_PASS1         "xxxx"     // [Password1] Wifi password

#define USE_EXPRESSION         // Add support for expression evaluation in rules (+3k2 code, +64 bytes mem)  
#define SUPPORT_IF_STATEMENT   // Add support for IF statement in rules (+4k2 code, -332 bytes mem)  

#define MQTT_TOPIC             "%12X"   // [Topic] unique MQTT device topic including (part of) device MAC address
#define MQTT_HOST              "xxxxx.s1.eu.hivemq.cloud"                // [MqttHost]
//#define MQTT_PORT              1883              // [MqttPort] MQTT port (10123 on CloudMQTT)
#define MQTT_PORT              8883              // [MqttPort] MQTT port (10123 on CloudMQTT)
#define MQTT_USER              "xxxxx"       // [MqttUser] MQTT user
#define MQTT_PASS              "xxxxx"       // [MqttPassword] MQTT password
#define TELE_PERIOD            300                // [TelePeriod] Telemetry (0 = disable, 10 - 3600 seconds)
#define SUPPORT_MQTT_EVENT
#define MQTT_TELE_RETAIN       1                 // Tele messages may send retain flag (0 = off, 1 = on)

#define USE_MCP230xx_ADDR 0x20
#define USE_MCP230xx                 // Enable INPUT mode (pinmode 1 through 4)
#define USE_MCP230xx_OUTPUT          // Enable OUTPUT mode (pinmode 5)
#define USE_MCP230xx_DISPLAYOUTPUT   // Display state of OUTPUT pins on main Tasmota web interface

#define MQTT_FINGERPRINT1     "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00"
#define USE_MQTT_TLS                             // Use TLS for MQTT connection (+34.5k code, +7.0k mem and +4.8k additional during connection handshake)
#define MQTT_TLS_ENABLED       true              // [SetOption103] Enable TLS mode (requires TLS version)

#define USE_RC_SWITCH         // Add support for RF transceiver using library RcSwitch (+2k7 code, 460 iram)

#define USE_DISPLAY
#define USE_DISPLAY_TM1637

#define USE_TM1638 // Add support for TM1638 switches copying Switch1 .. Switch8 (+1k code)
//#define TM1638_USE_AS_BUTTON // Add support for buttons
#define TM1638_USE_AS_SWITCH // Add support for switches (default)
#define TM1638_MAX_DISPLAYS 8 // Add support for power control 8 displays
#define TM1638_MAX_KEYS 8 // Add support for 8 keys
#define TM1638_MAX_LEDS 8 // Add support for 8 leds 

One more thing i am already using ESP32 with tasmota firmware instaalled using web installer and then i custom compiled the development release and did not reset old configuration , i only configure the pins for TM1638 and used it and faced the said issue.

Is it required to reset configuration like all parameter wifi ssid, mqtt setting etc 
After compiling i got 2 files tasmota32.bin and tasmota32.factory.bin, i used tasmota firmware update web page and burned tasmota32.bin file. 

![image](https://user-images.githubusercontent.com/29135632/223319777-c52afb3c-c0a0-4c49-9c13-95bac0d05a5f.png)

![Untitled](https://user-images.githubusercontent.com/29135632/223319949-2fb359dd-9210-40c6-8af8-f9a224324479.png)

Log is as follows

00:00:00.003 HDW: ESP32-D0WDQ6 
00:00:00.182 UFS: FlashFS mounted with 268 kB free
00:00:00.221 CFG: Loaded from File, Count 1062
00:00:00.672 BRY: Berry initialized, RAM used=4073 bytes
00:00:00.696 TFS: File '.drvset003' not found
00:00:00.697 CFG: Energy use defaults as file system not ready or file not found
00:00:00.701 DSP: TM1638 with 8 digits (type 0)
00:00:00.809 Project tasmota - Pump controller Version 12.4.0.2(tasmota32)-2_0_7(2023-03-02T14:35:42)
00:00:01.248 RUL: SYSTEM#INIT performs "backlog event reset"
00:00:01.293 RSL: RESULT = {"Event":"Done"}
00:00:01.315 RUL: EVENT#RESET performs "backlog ruletimer1 0; displaymode 1; var1 1; var2 ""
00:00:01.793 WIF: Connecting to AP1 USR1 Channel 4 BSSId EC:AD:E0:38:B0:D0 in mode 11n as 18FE34CB2D3B-8072...
00:00:02.001 RSL: RESULT = {"T1":0,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
00:00:02.260 RSL: RESULT = {"DisplayMode":1}
00:00:02.459 RSL: RESULT = {"Var1":"1"}
00:00:03.411 RSL: RESULT = {"Var2":""}
00:00:03.442 WIF: Connected
00:00:03.544 HTP: Web server active on 18FE34CB2D3B-8072 with IP address 192.168.0.101
00:00:03.551 UPP: Multicast (re)joined
09:58:20.556 MQT: Attempting connection...
09:58:23.330 MQT: TLS connected in 2766 ms, stack low mark 1916
09:58:23.332 MQT: MFLN not supported by TLS server
09:58:23.332 MQT: Connected
09:58:23.339 MQT: tele/18FE34CB2D3B/LWT = Online (retained)
09:58:23.342 MQT: cmnd/18FE34CB2D3B/POWER = 
09:58:23.354 MQT: tele/18FE34CB2D3B/INFO1 = {"Info1":{"Module":"ESP32-DevKit","Version":"12.4.0.2(tasmota32)","FallbackTopic":"cmnd/DVES_B49F88_fb/","GroupTopic":"cmnd/tasmotas/"}}
09:58:23.384 MQT: tele/18FE34CB2D3B/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"18FE34CB2D3B-8072","IPAddress":"192.168.0.101","IP6Global":"","IP6Local":"fe80::266f:28ff:feb4:9f88"}}
09:58:23.412 MQT: tele/18FE34CB2D3B/INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":117}}
09:58:23.429 MQT: stat/18FE34CB2D3B/RESULT = {"POWER1":"OFF"}
09:58:23.432 MQT: stat/18FE34CB2D3B/POWER1 = OFF
09:58:23.436 MQT: stat/18FE34CB2D3B/RESULT = {"POWER2":"OFF"}
09:58:23.440 MQT: stat/18FE34CB2D3B/POWER2 = OFF
09:58:23.444 MQT: stat/18FE34CB2D3B/RESULT = {"POWER3":"OFF"}
09:58:23.448 MQT: stat/18FE34CB2D3B/POWER3 = OFF
09:58:23.456 MQT: stat/18FE34CB2D3B/RESULT = {"POWER4":"OFF"}
09:58:23.460 MQT: stat/18FE34CB2D3B/POWER4 = OFF
09:58:23.464 MQT: stat/18FE34CB2D3B/RESULT = {"POWER5":"OFF"}
09:58:23.468 MQT: stat/18FE34CB2D3B/POWER5 = OFF
09:58:23.472 MQT: stat/18FE34CB2D3B/RESULT = {"POWER6":"OFF"}
09:58:23.486 MQT: stat/18FE34CB2D3B/POWER6 = OFF
09:58:23.490 MQT: stat/18FE34CB2D3B/RESULT = {"POWER7":"ON"}
09:58:23.493 MQT: stat/18FE34CB2D3B/POWER7 = ON
09:58:24.096 MQT: stat/18FE34CB2D3B/RESULT = {"Br":"true"}
09:58:24.568 MQT: stat/18FE34CB2D3B/RESULT = {"Br":"true"}
09:58:25.049 MQT: stat/18FE34CB2D3B/RESULT = {"Br":"true"}
09:58:25.493 MQT: stat/18FE34CB2D3B/RESULT = {"Br":"true"}
09:58:25.523 UPP: Multicast (re)joined
09:58:25.631 MQT: stat/18FE34CB2D3B/RESULT = 4D7944726976657220696E69746961 [MyDriver initial] ...
09:58:25.636 MQT: stat/18FE34CB2D3B/RESULT = {"Br":"nil"}
09:58:25.677 MQT: stat/18FE34CB2D3B/RESULT = {"POWER2":"ON"}
09:58:25.682 MQT: stat/18FE34CB2D3B/POWER2 = ON
09:58:25.820 MQT: stat/18FE34CB2D3B/RESULT = {"POWER2":"OFF"}
09:58:25.824 MQT: stat/18FE34CB2D3B/POWER2 = OFF
09:58:26.035 MQT: stat/18FE34CB2D3B/RESULT = {"Br":"nil"}
09:58:26.347 MQT: stat/18FE34CB2D3B/RESULT = {"Br":"nil"}
09:58:26.470 MQT: stat/18FE34CB2D3B/RESULT = {"Br":"nil"}
09:58:26.653 MQT: stat/18FE34CB2D3B/RESULT = 636C61737320477265657465722069 [class Greeter in] ...
09:58:26.658 MQT: stat/18FE34CB2D3B/RESULT = {"Br":"nil"}
09:58:26.794 MQT: stat/18FE34CB2D3B/RESULT = 636C61737320477265657465722069 [class Greeter in] ...
09:58:26.799 MQT: stat/18FE34CB2D3B/RESULT = {"Br":"nil"}
09:58:27.566 MQT: tele/18FE34CB2D3B/STATE = {"Time":"2023-03-07T09:58:27","Uptime":"0T00:00:11","UptimeSec":11,"Heap":115,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":230,"MqttCount":1,"Berry":{"HeapUsed":11,"Objects":171},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","POWER5":"OFF","POWER6":"OFF","Dimmer":88,"Color":"E08682","HSBColor":"2,42,88","Channel":[88,53,51],"Scheme":0,"Width":1,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"USR1","BSSId":"EC:AD:E0:38:B0:D0","Channel":4,"Mode":"11n","RSSI":100,"Signal":-38,"LinkCount":1,"Downtime":"0T00:00:03"}} (retained)
09:58:27.630 MQT: tele/18FE34CB2D3B/SENSOR = {"Time":"2023-03-07T09:58:27","Switch1":"ON","Switch2":"OFF","nil":{"AX":10,"AY":20,"AZ":30},"18FE34CB2D3B":{"Dimmer1Speed":0,"Dimmer2Speed":0,"Dimmer3Speed":0,"Humidity":1,"SwitchStatus":64,"TemperatureC":2,"TemperatureF":3},"TempUnit":"C"} (retained)
09:58:28.347 MQT: tele/18FE34CB2D3B/RESULT = {"Time":"2023-03-07T09:58:28","RfReceived":{"Data":"0x2DBEFB6","Bits":25,"Protocol":11,"Pulse":488}}



10:15:52.996 CMD: gpio
10:15:53.015 MQT: stat/18FE34CB2D3B/RESULT = {"GPIO0":{"66":"Button_n3"},"GPIO1":{"0":"None"},"GPIO2":{"224":"Relay1"},"GPIO3":{"0":"None"},"GPIO4":{"480":"Buzzer"},"GPIO5":{"1376":"WS28121"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"1056":"IRsend1"},"GPIO13":{"225":"Relay2"},"GPIO14":{"7648":"Input1"},"GPIO15":{"160":"Switch1"},"GPIO16":{"2048":"TM1638 CLK"},"GPIO17":{"2080":"TM1638 DIO"},"GPIO18":{"228":"Relay5"},"GPIO19":{"229":"Relay6"},"GPIO21":{"640":"I2C SDA1"},"GPIO22":{"608":"I2C SCL1"},"GPIO23":{"230":"Relay7"},"GPIO25":{"2112":"TM1638 STB"},"GPIO26":{"0":"None"},"GPIO27":{"1152":"RFrecv"},"GPIO32":{"0":"None"},"GPIO33":{"0":"None"},"GPIO34":{"161":"Switch2"},"GPIO35":{"0":"None"},"GPIO36":{"0":"None"},"GPIO39":{"0":"None"}}



it is not at all working when only the TM1638 GPIO are configured
i tried following command individually and with backlog too

Backlog DisplayModel 15; DisplayWidth 8; DisplayType 0; DisplayMode 0; DisplayDimmer 100; DisplayFloat 12.345678

**Reply by barbudor** [reply]
Hi @usrrsr
This is confirming what we discovered yesterday with another use case.

In the current code there is a conflict between the WS2812 driver (actually with the base class Light driver from which Ws2812 is a child) and other drivers including the Led/Button driver of TM1638.

As this is a device already in production, you will have to wait until the issue is fixed which won't happen until next week.
In the meantime the only immediate work around is to have an esp8266/32 dedicated to the Tm1638.

**Reply by usrrsr** [reply]
Well can i disable ws2812 driver to enable tm1638, is there any way out to use 7 segment ,leds and switches of  Tm1638 with disabling ws2812 pin assignment

**Reply by barbudor** [reply]
If you can disable the Ws2812, then the Tm1638 should immediately work

**Comment by usrrsr** [comment]
I tried to do so on my esp32 ie when i configure earlier ws2812 configured pin gpio05 as switch ,none ,button ,dht11 or something else my configuration of io gets reset 
This happened multiple time so finally i loose the hope

**Reply by barbudor** [reply]
I will try again tonight with your exact configuration
I was expecting only WS2812 to be the problem but may be other are
Nevertheless, this is a bug that will be solved next week
So don't loose hope, just be patient

thanks

**Comment by usrrsr** [comment]
Attached is my compiled file
I am using following board
ESP-WROOM-32D
https://a.aliexpress.com/_mOPK6bc

**Reply by barbudor** [reply]
Your problem is the non-continugous numbers, especially the Relays
- You have 1 button => It should be Button_n_1 (although is this case, it would work because you have no other buttons as your TM1638 is configured as Switches)
- Swiches 1 and 2 are ok => the TM1638 will be Swich3 to Switch10
- Relays 1, 2, 5, 6, 7 => There is no Relay 3 nor Relay 4 => This is forbidden, you need to have Relay 3 and 4. Either you shift you other relays to Relay 1..Relay 5, or you create dummy Relay 3 and Relay 4 on unused GPIOs
If you use Relay 1..Relay 5, then the leds of the TM1638 will be Power6 to Power13, Power14 is the Display on/off

Using continuous number is a mandatory requirement, especially when using virtual Switches/Buttons and relays

**Reply by usrrsr** [reply]
Oh
Thanks for correcting, i will check today evening and revert back

**Reply by usrrsr** [reply]
Hi i set configuration as below
![Configuration](https://user-images.githubusercontent.com/29135632/223750389-f6208ce3-6d73-4166-a0ad-6374da9994f6.jpg)
when i changes IO GPIO5 from WS2812 to None or Relay boards get reset and pin configuration get lost , all pin gets configured as None.

Is it that TM1638 pins should be disabled along with WS2812 and the reconfigured ?

I am badly waiting for an update to use WS2812 and TM1638 or get this problem of reseting after GPIO05 setting to None resolved

**Reply by barbudor** [reply]
What you see seems to be typical of 2 cases:

1) a bootloop
Something is crashing which makes the ESP restart in loop and Tasmota decide to remove parts of the configuration until the crash stops
Typically you should have around the begginning of the logs for the last boot something like "FRC: Some settings has been reset"
https://tasmota.github.io/docs/FAQ/#frequent-rebootsbootloops

2) A button mis-configured like "Button 1" on a pin which has a pull-down or "Button_i 1" on a pin with a pull up. Tasmota interpret that as a recovery request and erase all configuration https://tasmota.github.io/docs/Device-Recovery/#recovery-techniques

It is difficult to tell which one is your case from the limited details you provided.
As you have coded your wifi credentials in the binary, it will always reconnect to your network

You have to carefully ready the logs from the reboot until the 1st Teleperiod message. There should be some clues here


So it looks like you have something wrong.

**Reply by usrrsr** [reply]
Thanks for support
i have already configure  the board with
SetOption65 1
SetOption36 0

If i disable WS2812 and TM1638 board wont restart 
later I tried following configuration only leaving other pin as None and also as per old configuration cmbination
GPIO33 : TM1638_CLK
GPIO26 : TM1638_DIO
GPIO32 : TM1638_STB
but whenever i configure it for TM1638 boards ressets and keep on rebooting so final option is only to connect it on serial port install wit web installer and then upload custom binary file, its a very time consuming process so i think there should be one option in web installer to install custom local compiled binary file to reduce time cycle

As far as TM1638 is concerned i have really gived up, i will try only after tasmota get resolved WS2812 and TM1638 issue

But you are really helpful and help me out to troubleshoot the TM1638

Thanks again

**Reply by usrrsr** [reply]
Hi
Finally it worked, i did complete reconfiguration , disabled ws2812 and all switches
Now TM1638 is working and reliable too

Thanks for your support

**Reply by barbudor** [reply]
The problem has been fixed
Using the below configuration with:
- Switches 1 to 3
- Relay 1 to 5
- 1 WS2812

Then
- The GPIO Switches are Switch1 to Switch33
- The TM1638 buttons are Switch4 to Switch11
- The Relays are POWER1 to POWER5
- The TM1638 LEDs are POWER6 to POWER13
- The WS2812 is POWER14
- The TM1638 7-segment is POWER15

![image](https://user-images.githubusercontent.com/2996042/224549376-9068835f-3f67-4057-865a-d5c190f2e2a1.png)

![image](https://user-images.githubusercontent.com/2996042/224549692-71b2184d-d32c-417a-9db6-1d6234ac5206.png)

**Reply by usrrsr** [reply]
Do i need to update firmware or just do reconfiguration 
Can i add irsend, rfrecv in addition to ws2812

Gpio27 is configured as reset , is it required

**Reply by barbudor** [reply]
Of course you need to pull the update source code and recompile with your user_config_override.h. The changes won't magically jump into your esp.

IR should work too

**Reply by usrrsr** [reply]
Thanks, i will do the same

**Reply by usrrsr** [reply]
Hello,
I tested the latest build from development branch with custom option, 
I herewith confirm that board is working perfectly fine without any issue

Thanks for support
